spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/datastore/GcpDatastoreAutoConfiguration.java

```diff
@@ -101,21 +106,34 @@
 	}
 
 	@Bean
-	@ConditionalOnMissingBean
-	public Datastore datastore() {
-		DatastoreOptions.Builder builder = DatastoreOptions.newBuilder()
-				.setProjectId(this.projectId)
-				.setHeaderProvider(new UserAgentHeaderProvider(this.getClass()))
-				.setCredentials(this.credentials);
-		if (this.namespace != null) {
-			builder.setNamespace(this.namespace);
-		}
-
-		if (this.host != null) {
-			builder.setHost(this.host);
+	@ConditionalOnMissingBean(value = Datastore.class, parameterizedContainer = Supplier.class)
+	public Supplier<Datastore> datastoreSupplier(
+			ObjectProvider<DatastoreNamespaceProvider> namespaceProvider,
+			ObjectProvider<Datastore> datastore) {
+		if (datastore.getIfAvailable() != null) {
+			namespaceProvider.ifAvailable(unused -> {
+				throw new DatastoreDataException(
+						"A Datastore namespace provider and Datastore client were both configured. " +
+								"Only one can be configured.");
+			});
+			return () -> datastore.getIfAvailable();
 		}
-
-		return builder.build().getService();
+		return new CachingDatastoreProvider<>(
+				namespaceProvider.getIfAvailable() == null ? () -> this.namespace : namespaceProvider.getIfAvailable(),
+				namespace -> {
+			DatastoreOptions.Builder builder = DatastoreOptions.newBuilder()
```

Comment:
Not exactly true. This logic here *only deals with utilizing a custom namespace* whereas the `CachingDatastoreProvider` can be used with any combination of input settings. What nails `CachingDatastoreProvider` to Datastore is the `Datastore` type.

Reply:
I gave this try and there is a sticking point.

there are references to `this.projectid, this.credentials, this.host`. These three things felt strange to pass into the `CachingDatastoreProvider` since it lives in the Spring Data Datastore module instead of in autoconfig here. 

I don't want to  move `CachingDatastoreProvider` to the auto-config module because its behavior is something I want to verify in DatastoreTemplateTests (I have a unit test there that makes sure the correct clients are being used by template operations).